### PR TITLE
Extract reporter data generation, and use pdata

### PR DESCRIPTION
**Pull request by dmathieu**
This is a proposal to help reduce the code repetition in https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/208
It extracts the data generation into its own internal module, and switches from OTLP to pdata (which is an OTLP abstraction), so both reporters can rely on the same data generator.

### CODE REVIEW

**Comment by dmathieu** [comment]
This is blocked on https://github.com/open-telemetry/opentelemetry-collector/pull/11706, since pdata currently uses a map for the attribute table.

**Comment by dmathieu** [comment]
@open-telemetry/ebpf-profiler-approvers this change would greatly reduce the size of the OTLP and Collector reporters, as both would use pdata rather than OTLP and pdata.

File: reporter/internal/pdata/generate.go
Comment (florianl) [inline reporter/internal/pdata/generate.go:114]: Missed fix from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/247:
```suggestion
					ei, exists := p.Executables.GetAndRefresh(traceInfo.Files[i], ...)
```

File: reporter/otlp_reporter.go
Comment (florianl) [inline reporter/otlp_reporter.go:204]: Fix from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/256:
```suggestion
	if frameMapLock, exists := r.pdata.Frames.GetAndRefresh(frameID.FileID(), ...); exists {
```

File: reporter/internal/pdata/generate.go
Comment (florianl) [inline reporter/internal/pdata/generate.go:153]: Fix from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/256:
```suggestion
				fileIDInfoLock, exists := p.Frames.GetAndRefresh(traceInfo.Files[i], ...)
```

File: reporter/otlp_reporter.go
Comment (florianl) [inline reporter/otlp_reporter.go:187]: Fix from https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/256:
```suggestion
	_, known := r.pdata.Executables.GetAndRefresh(fileID, ...)
```

**Review by florianl — COMMENTED** [review]
Just some first spots. At the moment it is hard to review, as there is no backend that accepts this data. I will try to provide a devfiler version supporting v1.4.0 to continue here.

File: reporter/otlp_reporter.go
Comment (dmathieu) [reply reporter/otlp_reporter.go]: These are rather misleading, because your PR isn't merged yet ...

**Comment by florianl** [comment]
Here is a version of devfiler with OTel Profiling v1.4.0 support:

```
curl -L -H 'Authorization: ca14712d70ea6f6e' -o 'devfiler-v0.9.1-beta-v1.4.0.tar.gz' https://upload.elastic.co/d/b55a069766c48b8c0d4c0f2b854b0e846ee4deaea75596b68a8e1d88172dd0ca
```
or

https://upload.elastic.co/d/b55a069766c48b8c0d4c0f2b854b0e846ee4deaea75596b68a8e1d88172dd0ca
Authentication token: `ca14712d70ea6f6e`

**Comment by dmathieu** [comment]
😢 moving this back to draft. The changes in #237 are causing a lot of conflicts.

File: reporter/internal/pdata/pdata.go
Comment (christos68k) [inline reporter/internal/pdata/pdata.go:17]: We don't need this as it's already defined in `generate.go`

```suggestion
```

File: reporter/internal/pdata/pdata.go
Comment (christos68k) [inline reporter/internal/pdata/pdata.go:55]:
```suggestion
	frames.SetLifetime(FramesCacheLifetime) // Allow GC to clean stale items.
```

File: reporter/internal/pdata/pdata.go
Comment (christos68k) [inline reporter/internal/pdata/pdata.go:47]:
```suggestion
	executables.SetLifetime(ExecutableCacheLifetime) // Allow GC to clean stale items.
```

**Review by christos68k — COMMENTED** [review]
Did a pass, looks good to me. I'll spend a little more time with `generate.go` tomorrow.

File: reporter/internal/pdata/generate.go
Comment (florianl) [inline reporter/internal/pdata/generate.go:192]: This is incorrect, as frames of type `libpf.AbortFrame` are skipped. The correct value would be the number of interpreted frames + number of native frames.
But I think this is also a bug in the current implementation.

File: reporter/internal/pdata/generate.go
Comment (florianl) [inline reporter/internal/pdata/generate.go:126]: I'm wondering where this function is coming from and why it is called the way it is. `Strindex` should be `StrIndex`.
```suggestion
					mapping.SetFilenameStrIndex(getStringMapIndex(stringMap, fileName))
```

File: reporter/internal/pdata/pdata.go
Comment (florianl) [inline reporter/internal/pdata/pdata.go:34]: As all other arguments are also written in their extended form, `sps` should also be `samplesPerSecond`. WIthout implementation knowledges, `sps` could be anything otherwise.
```suggestion
func New(samplesPerSecond int, executablesCacheElements, framesCacheElements uint32,
```

**Review by florianl — COMMENTED** [review]
Just a first impression. Will have a deeper look again.

File: reporter/internal/pdata/generate.go
Comment (dmathieu) [reply reporter/internal/pdata/generate.go:126]: It comes from the pprofile package.
https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pprofile#Function.SetFilenameStrindex

That code is auto-generated from configuration to match the proto fields.
https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/cmd/pdatagen/internal/pprofile_package.go#L361
We could rename it in the code generator. For a full match, we'd have to rename the proto field as well though.

File: reporter/internal/pdata/generate.go
Comment (dmathieu) [reply reporter/internal/pdata/generate.go:192]: Yes, the main branch has the same logic.
https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/9eb454541a3cf1282a8755dbcec9a9471b243e83/reporter/otlp_reporter.go#L672

Let's open an external issue about this.

Comment (dmathieu) [reply reporter/internal/pdata/generate.go:192]: See https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/262

File: reporter/internal/pdata/generate.go
Comment (florianl) [inline reporter/internal/pdata/generate.go:55]: In the scope of message Profile the following attributes are missing:

- `default_sample_type_strindex` https://github.com/open-telemetry/opentelemetry-proto/blob/2080d87c780b5a8c65c9ae2771130fd669cef480/opentelemetry/proto/profiles/v1development/profiles.proto#L234C9-L234C37
- `period_type` https://github.com/open-telemetry/opentelemetry-proto/blob/2080d87c780b5a8c65c9ae2771130fd669cef480/opentelemetry/proto/profiles/v1development/profiles.proto#L223C13-L223C24

**Review by florianl — APPROVED** [review]
Just a minor last comment. Approving as merging this allows to unblock other work. Minor things can still be changed in subsequent steps.

File: reporter/internal/pdata/generate.go
Comment (dmathieu) [reply reporter/internal/pdata/generate.go:55]: I missed those in my collector upgrade. I'll add them there, thanks.

Comment (dmathieu) [reply reporter/internal/pdata/generate.go:55]: See https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/266

File: reporter/internal/pdata/generate.go
Comment (christos68k) [inline reporter/internal/pdata/generate.go:32]: Nit:

```suggestion
	sp := rp.ScopeProfiles().AppendEmpty()
```

**Review by christos68k — APPROVED** [review]

**Review by rockdaboot — APPROVED** [review]
LGTM

**Pull request merged**
